### DISCUSSION TITLE
Euphony test CMakeList

### DISCUSSION THREAD

**Discussion by YoungSeokHong** [euphony-io/euphony#50] [Knowledge Share]
혼자서 멋있게 #46 이슈를 해결해보고 싶었는데 역부족이었던 것 같습니다 ㅠㅠ
혹시 좋은 방법이 있으시면 공유해서 같이 해결해 보면 좋을 것 같아서 discussion을 만들어봤습니다!

<hr>

CMakeList를 처음 접해봤는데 저한테는 너무 낯설어서 이것 저것 찾아보고 출력해보면서 정리해보다가 CMakeList가 처음이신분들이 저말고도 계실 것 같아서 참고용으로 올립니다! 저도 잘 모르고 적은거라 많이 틀릴 수도 있어서 참고용으로만 봐주시면 될 것 같아요ㅠㅠ 
CMakeList 고수님들 있으시면 읽어봐주시고 잘못 된 부분 말씀 해주시면 감사하겠습니다!
```cmake
# 요구하는 최소 버전
cmake_minimum_required(VERSION 3.4.1)

# TEST_EUPHONY라는 변수에 'testEuphony'라는 값을 집어넣음
# 이 밑으로도 set함수는 어떠한 변수에 어떠한 값을 집어 넣는 용도로 사용
set(TEST_EUPHONY testEuphony)

# Include GoogleTest library / Gtest를 위해 라이브러리를 불러오는 과정
set (GOOGLETEST_ROOT ${ANDROID_NDK}/sources/third_party/googletest)
add_library (gtest STATIC ${GOOGLETEST_ROOT}/src/gtest_main.cc ${GOOGLETEST_ROOT}/src/gtest-all.cc)
target_include_directories (gtest PRIVATE ${GOOGLETEST_ROOT})
target_include_directories (gtest PUBLIC ${GOOGLETEST_ROOT}/include)

# TEST_EUPHONY 즉 testEuphony실행시 필요한 소스(파일)들의 목록
add_executable(
        ${TEST_EUPHONY}
        asciiCharsetTest.cpp
        base2Test.cpp
        base16Test.cpp
        defaultCharsetTest.cpp
        FFTTest.cpp
        FSKTest.cpp
        hexVectorTest.cpp
        packetTest.cpp
        packetBuilderTest.cpp
        packetErrorDetectorTest.cpp
        packetWithFSKTest.cpp
        waveTest.cpp
        waveBuilderTest.cpp
        waveRendererTest.cpp
)

# 빌드에 필요한 타켓(라이브러리, 외부 프로젝트, 실행 프로그램 등) 설정
target_link_libraries(${TEST_EUPHONY} PUBLIC euphony gtest)

# 밑에서 계속해서 사용하는 주소를 미리 설정해놓는 위치
# ANDROID_ABI는 arm64-v8a, x86 등 CPU/ABI와 관련된 여러 변수를 모아놓은 집합
set(TARGET_TEST_DIR /data/local/tmp/${TEST_EUPHONY}) # Directory on device to push tests.
set(TARGET_TEST_LIB_DIR ${TARGET_TEST_DIR}/${ANDROID_ABI})
set(LIBCPP_SHARED_PATH ${ANDROID_NDK}/sources/cxx-stl/llvm-libc++/libs/${ANDROID_ABI}/libc++_shared.so)

# message확인 위치 -> \euphony\euphony\.cxx\cmake\debug\(각 ANDROID_ABI)\cmake_server_log.txt
# 아래 두 줄은 원하는 위치에 집어 넣어 상황을 볼 때 용이함
# message("FINDFIND")
# message(${TARGET_TEST_LIB_DIR})

# (중요) 밑에서 계속 사용하는 ADB를 여기서 지정해줌
# 나중에 ADB를 커멘드 할 때 사용하기 때문에 ADB = cd C:\Users\(user name)\AppData\Local\Android\Sdk && adb 로 생각해두면 됨
get_target_property( OBOE_LIBRARY_PATH oboe::oboe IMPORTED_LOCATION)
find_program(ADB NAMES adb PATHS ${ANDROID_SDK_ROOT}/platform-tools) # Verified to be working on Linux.

# adb를 실행해서 나온 값을 ADB_DEVICE_ABI에 집어넣음
# 핸드폰을 연결하면 arm64-v8a, AVD를 연결하면 x86이 된다고 보면됨
# ADB_DEVICE_ABI_RESULT은 기기가 0개나 2개 이상이 연결 됐을 경우 1 정상적으로 1개의 기기만 연결됐을 경우 0
execute_process(COMMAND ${ADB} shell getprop ro.product.cpu.abi
        OUTPUT_VARIABLE ADB_DEVICE_ABI
        RESULT_VARIABLE ADB_DEVICE_ABI_RESULT)

# 기기연결이 잘못 되어있을 경우 메세지만 출력 후 gtest를 거치지 않음
if( ${ADB_DEVICE_ABI_RESULT} )
    message("NO ABI OR MORE THAN ONE DEVICE/EMULATOR")
else()
    string(STRIP ${ADB_DEVICE_ABI} ADB_DEVICE_ABI) # ADB_DEVICE_ABI의 공백을 제거해주는 작업
# (중요) 디바이스를 연결한 뒤 빌드 후, 제거하고 AVD를 연결하고 빌드를 하면 오류가 나는데 여기서 나는 것으로 예상됨
    if(${ANDROID_ABI} STREQUAL ${ADB_DEVICE_ABI}) # (매우중요)(추측) 연결되어 있는 기기에 대한 작업들만 시키기 위해 필요
         message("FINDFIND")
         message(${TARGET_TEST_LIB_DIR})
        # Prepare gtest for unit-test
        add_custom_command(TARGET ${TEST_EUPHONY} POST_BUILD # testEuphony빌드 이후에 실행할 것을 의미
                COMMAND ${ADB} shell mkdir -p ${TARGET_TEST_LIB_DIR} # TARGET_TEST_LIB_DIR 경로의 디렉토리 생성

                # Push necessary libraries -> 필요한 라이브러리들을 아까 만든 경로에 집어넣음
                # (중요) (추측) 이때 위의 if문 때문에 현재 연결된 디바이스의 운영체제에만 해당하는 디렉토리들이 넘어감
                COMMAND ${ADB} push $<TARGET_FILE:euphony> ${TARGET_TEST_LIB_DIR}/
                COMMAND ${ADB} push ${OBOE_LIBRARY_PATH} ${TARGET_TEST_LIB_DIR}/
                COMMAND ${ADB} push ${ANDROID_NDK}/sources/cxx-stl/llvm-libc++/libs/x86/libc++_shared.so ${TARGET_TEST_LIB_DIR}/ 

                # Push Euphony Test Binary
                COMMAND ${ADB} push $<TARGET_FILE:${TEST_EUPHONY}> ${TARGET_TEST_LIB_DIR}/
                COMMAND ${ADB} shell chmod 755 ${TARGET_TEST_LIB_DIR}/${TEST_EUPHONY} # 권한변경 -> 소유자(쓰기, 읽기, 실행 O)  그 외 사용자 (쓰기 X/ 읽기, 실행 O)
                )

        # Run gtest & get the result.
        # gtest를 실행시킴 TARGET_TEST_LIB_DIR을 같은 이름의 TARGET_TEST_LIB_DIR을 통해서 cmake.run.test.script에 전달
        add_custom_command(TARGET ${TEST_EUPHONY} POST_BUILD
                COMMAND ${CMAKE_COMMAND} -DADB:STRING="${ADB}" -DTARGET_TEST_LIB_DIR:STRING="${TARGET_TEST_LIB_DIR}" -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake.run.test.script
                )
    endif()
endif()
```
제 생각에는 문제가 발생하는 이유가 ADB와 if문이 처음 빌드할 때 한 번만 실행돼서 디바이스가 바뀌면 다시 if문을 실행시키지 못해서 이러한 문제가 발생하는 것 같습니다. 이 if문과 ADB도 add_custom_command처럼 빌드 이후에 디바이스가 바뀐걸 한 번 더 확인하고 실행해주면 될 것 같은데 아직 방법을 못찾았습니다ㅠㅠ 혹시 좋은 의견 있으시면 말씀해주세요!

추가적으로 이것저것 찾다가 안드로이드 스튜디오에서 구글테스트 하는 방법에 대한 깃헙 주소도 찾아서 올립니다!
https://github.com/Mr-Goldberg/android-studio-googletest/blob/master/README.md

**Comment by YoungSeokHong** [comment]
github에는 엄청 이쁘게 나오네요ㅠ 안드로이드 스튜디오에서는 죄다 초록색으로 나와서 구분하느라 나름 힘들었는데ㅠㅠ

**Reply by designe** [reply]
https://plugins.jetbrains.com/plugin/10089-cmake-simple-highlighter
이런걸 쓰시면 도움이 되실껍니다 :)

**Comment by designe** [comment]
https://github.com/euphony-io/euphony/blob/54966925eb8a78d72e47eb31acd03c270f4c9694/euphony/src/main/cpp/tests/CMakeLists.txt#L63

```cmake
execute_process(COMMAND ${ADB} shell LD_LIBRARY_PATH=${TARGET_TEST_LIB_DIR} ${TARGET_TEST_LIB_DIR}/testEuphony
                OUTPUT_VARIABLE GTEST_EUPHONY_OUTPUT
                RESULT_VARIABLE GTEST_RESULT
                ERROR_VARIABLE GTEST_ERROR_OUTPUT
                )

if( ${GTEST_RESULT} GREATER 0 )
    string(STRIP ${GTEST_EUPHONY_OUTPUT} GTEST_EUPHONY_OUTPUT)
    string(REPLACE "\n\n" "\r[NEXT]\r" GTEST_EUPHONY_OUTPUT ${GTEST_EUPHONY_OUTPUT})
    string(REPLACE "\n" "\r" GTEST_EUPHONY_OUTPUT ${GTEST_EUPHONY_OUTPUT})
    message(FATAL_ERROR "** Gtest Failure (${GTEST_RESULT}) **\r${GTEST_EUPHONY_OUTPUT}")
else()
    message("**** Gtest Success ****")
endif()
```

참고로 현재 gtest가 도는 방식은 add_custom_command 와 execute_process 의 장점을 섞은 사례라고 보면 됩니다 :)
add_custom_command는 @YoungSeokHong 님이 빌드 이후에 실행시킬 수 있다고 잘 설명을 써두셨고,
execute_process는 시점을 선택할 수 없는 대신 결과값을 변수에 담아 활용할 수 있습니다 :)
2개를 엮으면 시점과 결과값을 함께 볼 수 있습니다!

**Comment by YoungSeokHong** [comment]
### 안드로이드 build cache 삭제

미션 기한이 오늘까진데, 추석지내느라 잠시 소흘했네요 ㅠㅠ 지금 문제를 해결할 만한 방법을 하나 발견한 것 같아서 글 올립니다!

조사해본 결과 안드로이드는 기본적으로 build시간이 너무 오래 걸리는 문제를 막기 위해서 build시에 cache를 사용합니다. 그래서 현재 #46 이슈 같은 경우에도 이미 성공적인 빌드가 일어났기 때문에 디바이스가 변경됐음에도 이를 인지하지 못하고 계속해서 같은 방식으로만 build를 진행합니다.

>![image](https://user-images.githubusercontent.com/80624082/134214849-f9014b6b-64b1-46d5-80c7-6c5236851d53.png)
>#46 같은 문제가 발생할 때 build 시간이 매우 짧은 것에서 아이디어를 얻었습니다. 

그래서 캐시가 저장되는 저장소를 추적한 다음, cache를 직접 삭제 해보았습니다.
저의 경우에는 캐시의 주소가 C:\Users\YoungHong\.gradle\caches\transforms-2\files-2.1\2fc06640f22f0b2c2519ff4c360c6065 이었습니다. (2fc06640f22f0b2c2519ff4c360c6065는 프로젝트에 대한 해시값으로 생각됩니다)

상황은 이렇습니다.
* 기기가 없는 상태로 test오류가 없는 build 진행 -> 잘돌아감
* 그후, 안드로이드 기기를 연결한 상태로 build 진행 -> 잘돌아감(잘 돌아가면 안됨)
* 캐시 삭제
* 안드로이드 기기가 연결된 상태로 한 번 더 build 진행

그 결과 원하던 오류 메세지가 나왔습니다.
![image](https://user-images.githubusercontent.com/80624082/134216553-beff1549-3733-478e-815e-ac2e816f0c5b.png)

CMakeList의 명령어를 Script로 넣어서 빌드 이후에 진행하게도 해봤는데 이또한 기기를 변경할 때의 오류만 잡아줬을 뿐 #46 은 해결하지 못했습니다. 하지만, cache를 지우는 방법은 앞서 얘기한 문제와 #46도 해결하면서 .cxx디렉토리를 지우는 방식 보다 훨씬 빠른 속도로 진행됩니다.(.cxx 디렉토리를 재생성하지 않아서) 따라서 해결 방식은 아래 3개 정도로 볼 수 있겠습니다.

1. 매 build마다 cache파일을 지워주거나 rebuild를 한다.
2. gradle을 통해 build시 cache를 사용하지 못하도록 설정해 놓는다.
3. CMakeList로 오류 메세지가 나타나지 않는 경우(기기없이 build, build success)이후에 cache디렉토리를 지워주는 작업을 진행한다.

2번 방식이 제일 괜찮은 것 같아서 gradle.properties로 설정을 해봤는데 안됩니다 ㅠㅠ, 지금은 마음 접고 CMake로 디렉토리 지우는 방법을 공부하고 있습니다!

> 어찌하다보니 제 개인 연구노트(?)가 되버린 것 같은데 부담없이 피드백과 의견공유 부탁드립니다! 혼자서 다 찾아서 하자니 다른 것도 해야되는데 살짝 부담스럽네요 ㅠ

**Reply by designe** [reply]
넘잘하고있군요! :) 👍

**Comment by kuro11pow2** [comment]
와 감사합니다 저도 CMake 처음 접해서 눈앞이 캄캄했었는데 도움 많이 받았습니다 ㅠㅠ

**Comment by seohy02** [comment]
여러가지 방법 시도해보다가 질문이 생겼습니다!!
CMakeLists.txt 파일에 약간의 변화를 주고 저장 후 build를 하면 .cxx 폴더를 삭제하지 않아도 디바이스 연결 변경이 잘 인식되더라구요.

예를들면,
https://github.com/euphony-io/euphony/blob/54966925eb8a78d72e47eb31acd03c270f4c9694/euphony/src/main/cpp/tests/CMakeLists.txt#L38-L40

를 아래와 같이 바꾸고 저장해줍니다 ( 단순 줄바꿈) . 심지어 이 방법은 빌드 속도가 .cxx 폴더를 삭제하는 것보다 훨씬 빨랐어요!!

```
execute_process(
        COMMAND ${ADB} shell getprop ro.product.cpu.abi
        OUTPUT_VARIABLE ADB_DEVICE_ABI
        RESULT_VARIABLE ADB_DEVICE_ABI_RESULT)
```

Refresh Linked C++ Project 옵션도 사용해보고, 위에 써주신 내용 참고해서 캐시 삭제도 다양하게 시도해보고, add_custom_command로 디바이스 인식 시점을 변경시켜 보려고도 해봤는데 다 디바이스 변경을 인식하지 못했고, 넘 어렵네요! 위 방법은 어떤 원리로 적용되는 것일까요 ..... 😥 해결의 실마리가 되었으면 좋겠네요 ㅠㅠㅠㅠ

**Reply by YoungSeokHong** [reply]
저도 이 상황을 여러번 경험했는데 이 경우에는 코드가 변경되었기 때문에 build시 캐시를 이용하지 않아서 되는 것 같습니다.
혹시 $HOME\\.gradle\caches\transforms-2\files-2.1 디렉토리를 직접 지워보셨나요? 이 디렉토리를 직접 지워보시면 지금 일어나는 상황하고 같을 것 같습니다!

**Reply by seohy02** [reply]
@YoungSeokHong 네 맞아요!! 저 폴더를 지우면 잘 됐었어요!
말씀하신 코드 변경 후 build 시 이용하지 않는 캐시와 저 폴더 캐시가 동일한 캐시라고 봐도 될까요? 
 감사합니다~!

**Reply by YoungSeokHong** [reply]
네 여러 사이트 찾아봤는데도 gradle이 build시 사용하는 캐시 경로는 저거밖에 못 찾았습니다ㅠㅠ

**Reply by YoungSeokHong** [reply]
이거 폴더 안지우시고 대신에 밑에 있는 이름 같은 bin파일을 지워도 잘 되는 것 같아요!

**Reply by designe** [reply]
이게 어려운게 맞습니다. Android Studio의 gradle이 CMake를 제대로 관리해주지 못하고 있다는 방증이죠!!
여러분들이 찾고 계신 방법은 아직까지 그 누구도 시도해보지 않은 방법이기 때문에
그래서 더더욱 어려우실꺼에요!

하지만 해결한다면 여러분들이 만든 코드가 안드로이드 네이티브 유닛테스트를 동작시키는 최고의 방법이 될거라고 생각합니다!

**Reply by seohy02** [reply]
@YoungSeokHong @designe  감사합니다 !!! 더 알아볼게요! 💥💥

**Comment by Jsueeee** [comment]
혹시 테스트 잘 안되시는 분들 gradle 버전 몇 쓰시나요??
저는 4.1.1 사용 중인데 캐시를 안 건드려도 621로 바꾸면 build failed 잘 나오고 61로 다시 바꾸면 gtest success 나오고 이상없이 잘 돌아가는 것 같습니다.

**Reply by YoungSeokHong** [reply]
우선 처음에 .cxx 디렉토리를 제거하고, 디바이스 연결을 안한 상태로 621로 바꾸고build가 오류없이 잘 되는 것을 확인하고 난 뒤에, 디바이스를 연결하시고 다시 build를 돌렸을 때 에러문이 뜨시나요?

**Reply by Jsueeee** [reply]
![image](https://user-images.githubusercontent.com/66951780/134636176-9daf8c2b-0d7d-4e9a-bec0-5b3c1d8b8cfd.png)

**Reply by Jsueeee** [reply]
그러면 .cxx 파일 지우시고 터미널에서 명령어를 gradlew build --refresh-dependencies 이걸 사용해서 빌드해보실 수 있을까요??

**Reply by Jsueeee** [reply]
1. .cxx 파일 지우기
2. gradlew build --refresh-dependencies 로 빌드하기( 621 로 바꾼 상태이면 사진처럼 build failed)
3. 그 다음부턴 .cxx 파일 안지우고 gradlew build 만 사용해서 빌드해도 621-failed , 61-success로 잘 나오는 것 같아요!

**Reply by YoungSeokHong** [reply]
네 이렇게 하면 당연히 dependecy를 초기화 하는거라 되긴 되는데, 우선 디바이스를 바꿀 때 시간도 쓸모없이 오래 걸리고 멘토님이 우리에게 요청하신건 코드를 변경해서 이런 설정없이 build만 돌렸을 때 문제가 해결되는 것이여서 코드를 수정하는 시도가 필요할 것 같습니다.
그 방법 중 하나로 찾은 것이 코드 내에서 캐시를 삭제해주는 것이고요.

**Reply by Jsueeee** [reply]
혹시 properties 에서 해보셨다는게  ``android.enableBuildCache=false`` 이거 말씀하시는건가요?

**Reply by YoungSeokHong** [reply]
네넵 맞습니다!

**Reply by Jsueeee** [reply]
아 저거 사용해도 빌드가 제대로 안되셨다는거죠??

**Reply by YoungSeokHong** [reply]
넵 빌드는 제대로 되는데 원하는 상황에서 오류문이 안떳습니다 ㅠ

**Reply by Jsueeee** [reply]
방금 저거 넣고 해봤는데
![image](https://user-images.githubusercontent.com/66951780/134640535-30ee9c55-2cdd-4346-af3f-b856ec8cf525.png)
이렇게 뜨면 오류문 나온 거 아닐까요ㅜㅜ?

**Reply by YoungSeokHong** [reply]
이게 처음 상태에서 디바이스 없이 빌드 먼저 돌리신 다음에 다시 디바이스 연결하신 뒤에 빌드를 돌리니 나온 오류문인가요?

**Reply by Jsueeee** [reply]
네 디바이스없이 돌리면
![image](https://user-images.githubusercontent.com/66951780/134640829-a84132a8-98b3-4a34-9853-75bf78adb904.png)
이렇게 나오네요

**Reply by YoungSeokHong** [reply]
없이 돌렸을 때도 오류문이 뜨시나요??

**Reply by Jsueeee** [reply]
아 다시 꽂고 돌렸는데 ㅋㅋㅋㅋ 다시 해볼게요

**Reply by YoungSeokHong** [reply]
그렇다면 이미 ADB_DEVICE_ABI 값이 arm64-v8a로 설정되어있는 것 같습니다!

**Reply by Jsueeee** [reply]
없이 돌렸을 때는 오류문 안뜨고 no devices 메세지 뜹니다!

**Reply by YoungSeokHong** [reply]
no devices 메세지 한 번만 캡쳐해주실 수 있나요? 죄송합니다 저는 그런 메세지가 안떴어서 ㅠㅠ

**Reply by Jsueeee** [reply]
> 네 디바이스없이 돌리면
> ![image](https://user-images.githubusercontent.com/66951780/134640829-a84132a8-98b3-4a34-9853-75bf78adb904.png)
> 이렇게 나오네요

앗 이런 메시지가 안뜨시나요??? 뭐지...!! 
이 사진이랑 똑같아요 no devices/emulators found
이 메세지가 나오는 경우는 621에 대한 메시지가 아예 안나오고, 
디바이스 연결하고 빌드하면 621이어서 안된다는 메시지가 나옵니다.
gradle.properties에 추가한 게 잘 되는 거 아닐까요?

**Reply by Jsueeee** [reply]
![image](https://user-images.githubusercontent.com/66951780/134642457-3cc187e5-9e45-4750-8c3f-07d3247b453a.png)

**Reply by YoungSeokHong** [reply]
저 사진이 조금 짤린것 같은데 fail이라고 뜬 것 아닌가요??

**Reply by Jsueeee** [reply]
아 success 가 떠야하나요??

**Reply by YoungSeokHong** [reply]
넵! 디바이스 연결이 안된 상태에서는
![image](https://user-images.githubusercontent.com/80624082/134643930-c9df3f56-cae1-4526-945d-835a6e77a9ff.png)
이 갈림길에서 ```ADB_DEVICE_ABI_RESULT```값이 1이어서 if문에 걸리고 gtest가 돌아가지 않아서 test코드가 에러가 있더라도 무시하고 success가 떠야 하는 걸로 알고 있습니다!

**Reply by Jsueeee** [reply]
아 복잡하네요 ㅠㅠ 밑에 올리신 것도 해봐야겠네요!!

**Reply by YoungSeokHong** [reply]
CMake에서 어떤 도구를 쓸 수 있는지를 모르니까 너무 어려운 것 같아요 ㅠㅠ 화이팅입니다!

**Comment by YoungSeokHong** [comment]
제가 말한 디렉토리를 CMakelist내에서 지워보려고 계속 시도해봤는데 자꾸 이상한 오류가 떠서 대신에 짝을 이루는 bin 파일(C:\\Users\\YoungHong\\.gradle\\caches\\transforms-2\\files-2.1\\2fc06640f22f0b2c2519ff4c360c6065.bin)을
```cmake
file(REMOVE ${BUILD_CACHE})
```
를 사용해서 지워봤더니 프로젝트가 원하던대로 잘 돌아갑니다.

```BUILD_CACHE```는 build.gradle(euphony)에서 
```gradle
externalNativeBuild {
            cmake {
                  ...
                  arguments "-DBUILD_CACHE=C:\\Users\\YoungHong\\.gradle\\caches\\transforms-2\\files-2.1\\2fc06640f22f0b2c2519ff4c360c6065.bin"
                  ...
            }
}
```
로 설정해줬습니다.
지금 바꾼 코드는 경로가 저래서 제 컴퓨터에서만 돌아가는 코드이고 저걸 ```${HOME}```과 같은 변수들을 써서 조금 더 일반화 시키고 싶은데 변수들이 정확히 어떤 것들이 있고 것들을 써야할 지를 모르겠어서 헤매고 있습니다. ㅠㅠ 아이디어 있으신 분들은 코멘트 부탁드립니다!

**Reply by seohy02** [reply]
정말 좋은 방법인 것 같아요.....👍👍👍 저도 저 방법으로 잘 되네요 !!!
제가 `project.gradle.gradleUserHomeDir` 이걸로 바꿔보았는데 잘 되는 것 같아요!!! 한번 확인 부탁드려요!
```
arguments "-DBUILD_CACHE=${project.gradle.gradleUserHomeDir}\\caches\\transforms-2\\files-2.1\\2fc06640f22f0b2c2519ff4c360c6065.bin"
```

**Reply by YoungSeokHong** [reply]
오 감사합니다!

**Reply by YoungSeokHong** [reply]
'2fc06640f22f0b2c2519ff4c360c6065' 이 부분도 조금 더 일반화 시켜보고 싶었는데 아직까지는 무리인 것 같네요ㅠ
진짜 감사합니다 딱 저걸 원하고 있었는데 덕분에 더 좋은 코드가 됐네요! 이 부분 정리해서 이슈에 올리겠습니다!
